# Gift Store Analysis
This is a transactional data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers. It is important to note that there negative values like returns and discounts.

In [1]:
%load_ext sql
%sql mysql+mysqldb://root:123@localhost/retail
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

In [17]:
%%sql
SELECT * from data LIMIT 5;

 * mysql+mysqldb://root:***@localhost/retail
5 rows affected.


[(536365, '85123A', 'WHITE HANGING HEART T-LIGHT HOLDER', 6, '12/1/2010 8:26', 2.55, 17850, 'United Kingdom'),
 (536365, '71053', 'WHITE METAL LANTERN', 6, '12/1/2010 8:26', 3.39, 17850, 'United Kingdom'),
 (536365, '84406B', 'CREAM CUPID HEARTS COAT HANGER', 8, '12/1/2010 8:26', 2.75, 17850, 'United Kingdom'),
 (536365, '84029G', 'KNITTED UNION FLAG HOT WATER BOTTLE', 6, '12/1/2010 8:26', 3.39, 17850, 'United Kingdom'),
 (536365, '84029E', 'RED WOOLLY HOTTIE WHITE HEART.', 6, '12/1/2010 8:26', 3.39, 17850, 'United Kingdom')]

## What is the total revenue generated by each country, and what is the percentage contribution of each country to the overall revenue?
This query uses a Common Table Expression (CTE) to calculate revenue (Quantity * UnitPrice) for each country. It filters out negative quantities to exclude returns or discounts from the calculation. A second CTE is then used to compute the total revenue across all countries. To determine each country's contribution to the overall revenue, the query performs a CROSS JOIN with the total revenue CTE. Finally, the ROUND function is applied to ensure the resulting percentage values are displayed with clean decimal places.

In [21]:
%%sql
WITH RevenueByCountry AS (
    SELECT 
        Country,
        ROUND(SUM(Quantity * UnitPrice), 2) AS TotalRevenue
    FROM data
    WHERE Quantity > 0 -- Exclude returns/discounts
    GROUP BY Country
),
TotalRevenue AS (
    SELECT SUM(TotalRevenue) AS OverallRevenue
    FROM RevenueByCountry
)
SELECT 
    r.Country,
    r.TotalRevenue,
    ROUND((r.TotalRevenue / t.OverallRevenue) * 100, 2) AS RevenuePercentage
FROM RevenueByCountry r
CROSS JOIN TotalRevenue t
ORDER BY r.TotalRevenue DESC;

 * mysql+mysqldb://root:***@localhost/retail
16 rows affected.


[('United Kingdom', 288389.23, 93.47),
 ('EIRE', 4337.23, 1.41),
 ('France', 4257.14, 1.38),
 ('Germany', 4242.29, 1.37),
 ('Norway', 1919.14, 0.62),
 ('Lithuania', 1598.06, 0.52),
 ('Iceland', 711.79, 0.23),
 ('Spain', 620.0, 0.2),
 ('Italy', 427.8, 0.14),
 ('Australia', 358.25, 0.12),
 ('Belgium', 346.1, 0.11),
 ('Japan', 320.08, 0.1),
 ('Switzerland', 303.4, 0.1),
 ('Portugal', 261.2, 0.08),
 ('Poland', 248.16, 0.08),
 ('Netherlands', 192.6, 0.06)]

Since the store is based in the UK, the majority of sales come from the UK. While several other European countries also contribute to sales, it's notable that Australia appears as a market although it is outside of Europe. Understanding revenue distribution by country helps prioritize markets for expansion or marketing efforts.

## How many customers have negative transactions (returns or discounts), and what is the impact on total revenue?
This query identifies transactions that either have negative quantities or are explicitly marked as “Discount.” It then aggregates the data by customer to count the number of negative transactions and calculate their total negative revenue impact. Finally, it computes the percentage of this negative revenue relative to the customer's total revenue, providing insight into how returns or discounts affect overall sales performance.

In [48]:
%%sql
WITH NegativeTransactions AS (
    SELECT 
        CustomerID,
        COUNT(DISTINCT InvoiceNo) AS NegativeTransactionCount,
        ROUND(SUM(Quantity * UnitPrice), 2) AS NegativeRevenue
    FROM data
    WHERE Quantity < 0 OR Description = 'Discount'
    GROUP BY CustomerID
),
TotalRevenue AS (
    SELECT ROUND(SUM(Quantity * UnitPrice), 2) AS OverallRevenue
    FROM data
)
SELECT 
    COUNT(DISTINCT n.CustomerID) AS CustomersWithNegatives,
    SUM(n.NegativeTransactionCount) AS TotalNegativeTransactions,
    ROUND(SUM(n.NegativeRevenue), 2) AS TotalNegativeRevenue,
    t.OverallRevenue,
    ROUND((SUM(n.NegativeRevenue) / t.OverallRevenue) * 100, 2) AS NegativeRevenuePercentage
FROM NegativeTransactions n
CROSS JOIN TotalRevenue t;

 * mysql+mysqldb://root:***@localhost/retail
1 rows affected.


[(62, Decimal('107'), -58978.61, 249553.86, -23.63)]

Returns and discounts account for a reduction of approximately one-quarter of the total sales. Understanding negative transactions from returns and discounts helps assess customer satisfaction and financial impact.

## Which products have sold the most units in total?
This query aggregates the total quantity sold for each product, explicitly filtering out any negative quantities to exclude returns or discounts. It then orders the results by total quantity in descending order and limits the output to the top five products, highlighting the best-selling items based on volume.

In [51]:
%%sql
SELECT 
    StockCode,
    Description,
    SUM(Quantity) AS TotalQuantity
FROM data
WHERE Quantity > 0
GROUP BY StockCode, Description
ORDER BY TotalQuantity DESC
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/retail
5 rows affected.


[('84077', 'WORLD WAR 2 GLIDERS ASSTD DESIGNS', Decimal('3465')),
 ('84950', 'ASSORTED COLOUR T-LIGHT HOLDER', Decimal('1866')),
 ('21915', 'RED  HARMONICA IN BOX ', Decimal('1779')),
 ('85123A', 'WHITE HANGING HEART T-LIGHT HOLDER', Decimal('1548')),
 ('17084R', 'ASSORTED INCENSE PACK', Decimal('1440'))]

The product 'WORLD WAR 2 GLIDERS ASSTD DESIGNS' outperforms other products, generating nearly twice the sales volume compared to the rest. Identifying top-selling products by quantity helps with inventory management.

## How many unique customers made purchases in each country?
This query counts the number of distinct CustomerID values per country, ensuring that only positive transactions are included by excluding any records with negative quantities (returns and discounts). It groups the data by Country and sorts the results by customer count, providing a clear view of customer distribution and activity across different regions.

In [55]:
%%sql
SELECT 
    Country,
    COUNT(DISTINCT CustomerID) AS UniqueCustomers
FROM data
WHERE Quantity > 0
GROUP BY Country
ORDER BY UniqueCustomers DESC;

 * mysql+mysqldb://root:***@localhost/retail
16 rows affected.


[('United Kingdom', 392),
 ('Germany', 9),
 ('France', 5),
 ('EIRE', 3),
 ('Portugal', 2),
 ('Australia', 1),
 ('Belgium', 1),
 ('Iceland', 1),
 ('Italy', 1),
 ('Japan', 1),
 ('Lithuania', 1),
 ('Netherlands', 1),
 ('Norway', 1),
 ('Poland', 1),
 ('Spain', 1),
 ('Switzerland', 1)]

Most customers are based in the UK, followed by Germany. In contrast, the majority of other countries have only a single customer each. Knowing the customer base by country informs marketing strategies.

## What is the average order value per invoice?
The subquery calculates the total order value for each invoice by summing the product of Quantity and UnitPrice, while filtering out negative quantities and using a HAVING clause to ensure only invoices with positive total values are included. The outer query then applies the AVG function to compute the average order value across all qualifying invoices. Finally, the ROUND function is used to present the result with two decimal places for clarity.

In [91]:
%%sql
SELECT 
    ROUND(AVG(OrderValue), 2) AS AvgOrderValue
FROM (
    SELECT 
        InvoiceNo,
        SUM(Quantity * UnitPrice) AS OrderValue
    FROM data
    WHERE Quantity > 0
    GROUP BY InvoiceNo
    HAVING SUM(Quantity * UnitPrice) > 0
) sub;

 * mysql+mysqldb://root:***@localhost/retail
1 rows affected.


[(510.82,)]

The average order value across all transactions is £510.82. This suggests that customers tend to place relatively large orders, which could be due to bulk buying behavior or the nature of the products being sold. Average order value helps assess customer spending behavior.